tcms/rpc/tests/test_testexecution.py

```diff
@@ -274,6 +276,78 @@ def test_get_links(self):
         self.assertEqual(logs[0]['url'], 'http://kiwitcms.org')
 
 
+class TestExecutionHistory(APITestCase):
+    def _fixture_setup(self):
+        super()._fixture_setup()
+
+        self.execution = TestExecutionFactory()
+
+    def test_history_for_non_existing_execution(self):
+        with self.assertRaisesRegex(XmlRPCFault, "TestExecution matching query does not exist."):
+            self.rpc_client.TestExecution.history(-5)
+
+    def test_history_new_execution(self):
+        execution = TestExecutionFactory()
+
+        history = self.rpc_client.TestExecution.history(execution.pk)
+
+        self.assertEqual(1, len(history))
+
+    def test_history(self):
+        """
+        Test that for an execution that has been updated 3 times,
+        there are 4 history entries (first one is the creation of the object).
+
+        Note: the `time.sleep` call after every update is necessary,
+        because otherwise the changes happen too fast,
+        and the measurable delta is 1 second.
```

Comment:
This shouldn't be the case. history_date is a DateTimeField which has sub-second resolution. Below is on sqlite: 

```
>>> for w in ["hello", 'loop', 'world']:
...     tp.text = w
...     tp.save()
... 
>>> for h in tp.history.all():
...     print(h.history_date.isoformat(), h)
... 
2020-09-30T12:48:13.957056 TP from Bot as of 2020-09-30 12:48:13.957056
2020-09-30T12:48:13.901642 TP from Bot as of 2020-09-30 12:48:13.901642
2020-09-30T12:48:13.699457 TP from Bot as of 2020-09-30 12:48:13.699457
2020-09-24T11:02:43.072550 TP from Bot as of 2020-09-24 11:02:43.072550
2020-09-18T13:16:48.548605 TP from Bot as of 2020-09-18 13:16:48.548605
```

The first 3 history entries have been updated in the for loop before and timestamp for all of them is different.

Reply:
```
for history_entry in history:
            print(history_entry['history_date'].isoformat())

Traceback (most recent call last):
  File "/Users/anton_sankov/Desktop/reno/git/kiwitcms/Kiwi/tcms/rpc/tests/test_testexecution.py", line 324, in test_history
    print(history_entry['history_date'].isoformat())
AttributeError: 'DateTime' object has no attribute 'isoformat'
````
I guess `.values()` changes the time of field and loses some precision. the test fails if i remove the sleep. will debug this later to see what exactly is happening.

Reply:
> I guess .values() changes the time of field and loses some precision

this isn't the case for me. 

```
>>> for obj in tp.history.all().order_by('-history_date').values():
...     print(obj['history_date'].isoformat(), type(obj['history_date']))
... 
2020-09-30T12:48:13.957056 <class 'datetime.datetime'>
2020-09-30T12:48:13.901642 <class 'datetime.datetime'>
2020-09-30T12:48:13.699457 <class 'datetime.datetime'>
2020-09-24T11:02:43.072550 <class 'datetime.datetime'>
2020-09-18T13:16:48.548605 <class 'datetime.datetime'>
```

what version of Python is this and which class exactly is the `DateTime` class referenced in your output? The standard class is all lowercase and what you are showing looks like a form field class, not a Python datetime object.

Reply:
I updated the doc string to acknowledge that

Reply:
Not a fault per-se but that's how the native Python xmlrpc client has been designed to work. It follows ISO 8601 which doesn't have sub-seconds precision. :-( https://en.wikipedia.org/wiki/ISO_8601.

Our json formatter doesn't suffer from this.